race condition breaks parallel run of `sc.datasets` functions (e.g. in tests)

### Please make sure these conditions are met

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of scanpy.
- [x] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

`_check_datafile_present_and_download` has a race condition:

e.g. `pbmc3k_processed` uses the pattern “check if a cached version exists in `sc.settings.datasetdir`. If not, download it to that location. If yes, read it”

[errors like this](https://github.com/lueckenlab/patpy/actions/runs/25002152698/job/73215072284?pr=126) means we have a corrupted file. I’m pretty sure that’s because of multiprocess pytest:
- `pytest-xdist` host process will figure out the list of tests and then distribute them between workers
- `pytest-xdist` worker 1 starts evaluating the fixture first, doesn’t see the cache file, and starts downloading the file
- `pytest-xdist` worker 2 starts evaluating the fixture too (even a session-scoped fixure is evaluated once per process), finds the partially-written file (without knowing it’s not done downloading), and tries to read it → error

so if I’m right, we could fix that on the scanpy side by either
- somehow setting a lock while writing, either manually (wait for lock instead of just doing `path.is_file()`) or by setting `locking=True` ([see here](https://docs.h5py.org/en/latest/high/file.html#h5py.File)). that’d mean that other processes wait until the one that first got there is done downloading.
- downloading to a temporary location and renaming the downloaded file to the cache location. this would cause parallel downloads but would also work since file renaming is an atomic operation

cc @VladimirShitov 

### Minimal code sample

```python
# /// script
# requires-python = ">=3.12"
# dependencies = [
#   "scanpy@git+https://github.com/scverse/scanpy.git@main",
# ]
# ///
#
# This script automatically imports the development branch of scanpy to check for issues

import scanpy as sc
# TODO: it’s late OK
```

### Error output

```pytb
0
_ ERROR at setup of TestCheckAdataLoaded.test_adata_loaded_true_after_prepare_anndata_wasserstein_tsne _
[gw1] linux -- Python 3.14.4 /home/runner/.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/bin/python3

    @pytest.fixture(scope="session")
    def pbmc3k_adata():
        """Preprocessed PBMC3k dataset with randomly assigned sample labels.
    
        Provides real single-cell data with X_pca embedding and louvain cell-type
        annotations, suitable for methods that require biological structure in the
        data (e.g. DiffusionEMD, GloScope, WassersteinTSNE, PILOT, MOFA).
        """
        import scanpy as sc
    
>       adata = sc.datasets.pbmc3k_processed()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

tests/conftest.py:89: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/scanpy/datasets/_utils.py:16: in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/scanpy/datasets/_datasets.py:452: in pbmc3k_processed
    return read(settings.datasetdir / "pbmc3k_processed.h5ad", backup_url=url)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/scanpy/readwrite.py:150: in read
    return _read(
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/scanpy/readwrite.py:838: in _read
    return read_h5ad(filename, backed=backed)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/anndata/_io/h5ad.py:263: in read_h5ad
    with h5py.File(filename, "r") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/h5py/_hl/files.py:555: in __init__
    fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../.local/share/hatch/env/virtual/patpy/NbKEWlp-/hatch-test.py3.14-pre/lib/python3.14/site-packages/h5py/_hl/files.py:232: in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OSError: Unable to synchronously open file (truncated file: eof = 3014656, sblock->base_addr = 0, stored_eof = 24653425)

h5py/h5f.pyx:106: OSError
```

### Versions

<details>

```
scanpy  1.13.0.dev61+g395006786
----    ----
pytz    2026.1.post1
zarr    3.1.6
typing_extensions       4.15.0
google-crc32c   1.8.0
kiwisolver      1.5.0
donfig  0.8.1.post1
MarkupSafe      3.0.3
PyYAML  6.0.3
threadpoolctl   3.6.0
packaging       26.1
Jinja2  3.1.6
pydantic-settings       2.14.0
dask    2026.3.0
pillow  12.2.0
coverage        7.13.5
python-dotenv   1.2.2
numpy   2.4.4
numcodecs       0.16.5
pydantic        2.13.3
scverse-misc    0.0.5
natsort 8.4.0
python-dateutil 2.9.0.post0
tblib   3.2.2
sparse  0.18.0
pyarrow 23.0.1
llvmlite        0.47.0
scipy   1.17.1
typing-inspection       0.4.2
six     1.17.0
legacy-api-wrap 1.5
anndata 0.12.10
toolz   1.1.0
pandas  2.3.3
session-info2   0.4.1
annotated-types 0.7.0
cloudpickle     3.1.2
matplotlib      3.10.8
scikit-learn    1.8.0
cycler  0.12.1
joblib  1.5.3
fsspec  2026.3.0
pydantic_core   2.46.3
fast-array-utils        1.4.1
msgpack 1.1.2
psutil  7.2.2
pyparsing       3.3.2
numba   0.65.0
h5py    3.16.0
----    ----
Python  3.14.4 (main, Apr  8 2026, 17:48:49) [GCC 15.2.1 20260209]
OS      Linux-6.19.13-arch1-1-x86_64-with-glibc2.43
CPU     16/16 logical CPU cores
Updated 2026-04-28 18:19
```

</details>

--- a/docs/release-notes/4142.fix.md
+++ b/docs/release-notes/4142.fix.md
@@ -1,1 +1,1 @@
-Download cached dataset files atomically so that parallel processes sharing a dataset directory (e.g. {mod}`pytest-xdist` workers) can no longer read a partially-written file {smaller}`gaoflow`
+Download cached dataset files atomically so that parallel processes sharing a dataset directory (e.g. `pytest-xdist` workers) can no longer read a partially-written file {smaller}`gaoflow`

--- a/docs/release-notes/4142.fix.md
+++ b/docs/release-notes/4142.fix.md
@@ -1,0 +1,1 @@
+Download cached dataset files atomically so that parallel processes sharing a dataset directory (e.g. {mod}`pytest-xdist` workers) can no longer read a partially-written file {smaller}`gaoflow`

--- a/src/scanpy/readwrite.py
+++ b/src/scanpy/readwrite.py
@@ -1093,10 +1093,7 @@ def _download(url: str, path: Path):
     blocksize = 1024 * 8
     blocknum = 0
 
-    # Download to a temporary file in the same directory and atomically move it
-    # into place once it is complete. This way concurrent readers (e.g. parallel
-    # pytest workers sharing a dataset cache) never observe ``path`` in a
-    # partially-written state. See #4097.
+    # Write to a temp file and rename so readers never see a partial file (#4097).
     tmp_path: Path | None = None
     try:
         req = Request(url, headers={"User-agent": "scanpy-user"})
@@ -1130,9 +1127,7 @@ def _download(url: str, path: Path):
         tmp_path = None
 
     except (KeyboardInterrupt, Exception):
-        # Make sure no half-downloaded temporary file is left behind. ``path``
-        # itself is only ever created by the atomic rename above, so it is left
-        # untouched (it may belong to another process that finished first).
+        # Only remove our own temp file; leave path, which may be another process's.
         if tmp_path is not None:
             tmp_path.unlink(missing_ok=True)
         raise

--- a/src/scanpy/readwrite.py
+++ b/src/scanpy/readwrite.py
@@ -1084,6 +1084,7 @@ def _get_filename_from_key(key, ext=None) -> Path:
 
 def _download(url: str, path: Path):
     from ssl import create_default_context
+    from tempfile import NamedTemporaryFile
     from urllib.request import Request, urlopen
 
     from certifi import contents
@@ -1092,6 +1093,11 @@ def _download(url: str, path: Path):
     blocksize = 1024 * 8
     blocknum = 0
 
+    # Download to a temporary file in the same directory and atomically move it
+    # into place once it is complete. This way concurrent readers (e.g. parallel
+    # pytest workers sharing a dataset cache) never observe ``path`` in a
+    # partially-written state. See #4097.
+    tmp_path: Path | None = None
     try:
         req = Request(url, headers={"User-agent": "scanpy-user"})
 
@@ -1105,8 +1111,14 @@ def _download(url: str, path: Path):
                     unit_divisor=1024,
                     total=total if total is None else int(total),
                 ) as t,
-                path.open("wb") as f,
+                NamedTemporaryFile(
+                    dir=path.parent,
+                    prefix=f"{path.name}.",
+                    suffix=".part",
+                    delete=False,
+                ) as f,
             ):
+                tmp_path = Path(f.name)
                 block = resp.read(blocksize)
                 while block:
                     f.write(block)
@@ -1114,10 +1126,15 @@ def _download(url: str, path: Path):
                     t.update(len(block))
                     block = resp.read(blocksize)
 
+        tmp_path.replace(path)
+        tmp_path = None
+
     except (KeyboardInterrupt, Exception):
-        # Make sure file doesn’t exist half-downloaded
-        if path.is_file():
-            path.unlink()
+        # Make sure no half-downloaded temporary file is left behind. ``path``
+        # itself is only ever created by the atomic rename above, so it is left
+        # untouched (it may belong to another process that finished first).
+        if tmp_path is not None:
+            tmp_path.unlink(missing_ok=True)
         raise
 
 

--- a/tests/test_datasets.py
+++ b/tests/test_datasets.py
@@ -228,7 +228,6 @@ def test_download_failure_keeps_existing_file(
     parallel-cache race could wipe a file another process had finished
     downloading. The atomic version only cleans up its own temporary file.
     """
-    import io
     import urllib.request
 
     from scanpy.readwrite import _download

--- a/tests/test_datasets.py
+++ b/tests/test_datasets.py
@@ -219,6 +219,48 @@ def test_download_atomic(tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> Non
     assert list(dest.parent.iterdir()) == [dest]
 
 
+def test_download_failure_keeps_existing_file(
+    tmp_path: Path, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """A failed download must not delete an already-present destination (#4097).
+
+    The old code unconditionally removed ``path`` on error, which under the
+    parallel-cache race could wipe a file another process had finished
+    downloading. The atomic version only cleans up its own temporary file.
+    """
+    import io
+    import urllib.request
+
+    from scanpy.readwrite import _download
+
+    dest = tmp_path / "cache" / "data.bin"
+    dest.parent.mkdir()
+    # A sibling process already finished this download.
+    dest.write_bytes(b"complete")
+
+    class FailingResponse:
+        def info(self) -> dict[str, str]:
+            return {"content-length": "100"}
+
+        def read(self, size: int) -> bytes:
+            raise OSError("connection reset")
+
+        def __enter__(self) -> Self:
+            return self
+
+        def __exit__(self, *exc: object) -> bool:
+            return False
+
+    monkeypatch.setattr(urllib.request, "urlopen", lambda *a, **k: FailingResponse())
+
+    with pytest.raises(OSError, match="connection reset"):
+        _download("http://example.invalid/data.bin", dest)
+
+    # The pre-existing file is untouched and no temporary file is left behind.
+    assert dest.read_bytes() == b"complete"
+    assert list(dest.parent.iterdir()) == [dest]
+
+
 # These are tested via doctest
 DS_INCLUDED = frozenset({"krumsiek11", "toggleswitch", "pbmc68k_reduced"})
 # These have parameters that affect shape and so on

--- a/tests/test_datasets.py
+++ b/tests/test_datasets.py
@@ -170,12 +170,7 @@ def test_download_failure() -> None:
 
 
 def test_download_atomic(tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> None:
-    """A download must be atomic (#4097).
-
-    The destination path must not appear until the download has completed, so
-    that concurrent readers (e.g. parallel pytest workers sharing a cache) never
-    observe a partially-written file.
-    """
+    """The destination must not appear until the download finished (#4097)."""
     import io
     import urllib.request
 
@@ -210,31 +205,21 @@ def test_download_atomic(tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> Non
     _download("http://example.invalid/data.bin", dest)
 
     assert dest.read_bytes() == content
-    # sanity check that the download was actually streamed in several chunks
     assert len(dest_present_during_download) > 1
-    assert not any(dest_present_during_download), (
-        "destination appeared before the download finished"
-    )
-    # the temporary file must have been renamed, leaving only the final file
+    assert not any(dest_present_during_download)
     assert list(dest.parent.iterdir()) == [dest]
 
 
 def test_download_failure_keeps_existing_file(
     tmp_path: Path, monkeypatch: pytest.MonkeyPatch
 ) -> None:
-    """A failed download must not delete an already-present destination (#4097).
-
-    The old code unconditionally removed ``path`` on error, which under the
-    parallel-cache race could wipe a file another process had finished
-    downloading. The atomic version only cleans up its own temporary file.
-    """
+    """A failed download must not delete an already-present destination (#4097)."""
     import urllib.request
 
     from scanpy.readwrite import _download
 
     dest = tmp_path / "cache" / "data.bin"
     dest.parent.mkdir()
-    # A sibling process already finished this download.
     dest.write_bytes(b"complete")
 
     class FailingResponse:
@@ -256,7 +241,6 @@ def test_download_failure_keeps_existing_file(
     with pytest.raises(OSError, match="connection reset"):
         _download("http://example.invalid/data.bin", dest)
 
-    # The pre-existing file is untouched and no temporary file is left behind.
     assert dest.read_bytes() == b"complete"
     assert list(dest.parent.iterdir()) == [dest]
 

--- a/tests/test_datasets.py
+++ b/tests/test_datasets.py
@@ -242,7 +242,8 @@ def test_download_failure_keeps_existing_file(
             return {"content-length": "100"}
 
         def read(self, size: int) -> bytes:
-            raise OSError("connection reset")
+            msg = "connection reset"
+            raise OSError(msg)
 
         def __enter__(self) -> Self:
             return self

--- a/tests/test_datasets.py
+++ b/tests/test_datasets.py
@@ -168,6 +168,56 @@ def test_download_failure() -> None:
     excinfo.value.close()
 
 
+def test_download_atomic(tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> None:
+    """A download must be atomic (#4097).
+
+    The destination path must not appear until the download has completed, so
+    that concurrent readers (e.g. parallel pytest workers sharing a cache) never
+    observe a partially-written file.
+    """
+    import io
+    import urllib.request
+
+    from scanpy.readwrite import _download
+
+    content = b"0123456789" * 5_000
+    dest = tmp_path / "cache" / "data.bin"
+    dest.parent.mkdir()
+    dest_present_during_download: list[bool] = []
+
+    class FakeResponse:
+        def __init__(self) -> None:
+            self._buf = io.BytesIO(content)
+
+        def info(self) -> dict[str, str]:
+            return {"content-length": str(len(content))}
+
+        def read(self, size: int) -> bytes:
+            chunk = self._buf.read(size)
+            if chunk:
+                dest_present_during_download.append(dest.exists())
+            return chunk
+
+        def __enter__(self) -> FakeResponse:
+            return self
+
+        def __exit__(self, *exc: object) -> bool:
+            return False
+
+    monkeypatch.setattr(urllib.request, "urlopen", lambda *a, **k: FakeResponse())
+
+    _download("http://example.invalid/data.bin", dest)
+
+    assert dest.read_bytes() == content
+    # sanity check that the download was actually streamed in several chunks
+    assert len(dest_present_during_download) > 1
+    assert not any(dest_present_during_download), (
+        "destination appeared before the download finished"
+    )
+    # the temporary file must have been renamed, leaving only the final file
+    assert list(dest.parent.iterdir()) == [dest]
+
+
 # These are tested via doctest
 DS_INCLUDED = frozenset({"krumsiek11", "toggleswitch", "pbmc68k_reduced"})
 # These have parameters that affect shape and so on

--- a/tests/test_datasets.py
+++ b/tests/test_datasets.py
@@ -21,6 +21,7 @@ from testing.scanpy._pytest.marks import needs
 
 if TYPE_CHECKING:
     from collections.abc import Callable
+    from typing import Self
 
     from anndata import AnnData
 
@@ -198,7 +199,7 @@ def test_download_atomic(tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> Non
                 dest_present_during_download.append(dest.exists())
             return chunk
 
-        def __enter__(self) -> FakeResponse:
+        def __enter__(self) -> Self:
             return self
 
         def __exit__(self, *exc: object) -> bool:

